examples/fun-token/src/lib.rs

```diff
@@ -0,0 +1,502 @@
+    }
+
+    pub fn total_balance(&self) -> Balance {
+        self.balance + self.locked_balances.values().sum::<Balance>()
+    }
+}
+
+#[near_bindgen]
+#[derive(BorshDeserialize, BorshSerialize)]
+pub struct FunToken {
+    /// AccountID -> Account details.
+    pub accounts: Map<AccountId, Account>,
+
+    /// Total supply of the all token.
+    pub total_supply: Balance,
+}
+
+impl Default for FunToken {
+    fn default() -> Self {
+        env::panic(b"Not initialized");
+        unreachable!();
+    }
+}
+
+#[near_bindgen(init => new)]
+impl FunToken {
+    pub fn new(owner_id: AccountId, total_supply: Balance) -> Self {
+        let mut ft = Self { accounts: Map::new(b"a".to_vec()), total_supply };
+        let mut account = ft.get_account(&owner_id);
+        account.balance = total_supply;
+        ft.accounts.insert(&owner_id, &account);
+        ft
+    }
+
+    /// Sets amount allowed to spent by `escrow_account_id` on behalf of the caller of the function
+    /// (`predecessor_id`) who is considered the balance owner to the new `allowance`.
+    /// If some amount of tokens is currently locked by the `escrow_account_id` the new allowance is
+    /// decreased by the amount of locked tokens.
+    pub fn set_allowance(&mut self, escrow_account_id: AccountId, allowance: Balance) {
+        let owner_id = env::predecessor_account_id();
+        if escrow_account_id == owner_id {
+            env::panic(b"Can't set allowance for yourself");
+        }
+        let mut account = self.get_account(&owner_id);
+        let locked_balance = account.get_locked_balance(&escrow_account_id);
+        if locked_balance > allowance {
+            env::panic(b"The new allowance can't be less than the amount of locked tokens");
+        }
+
+        account.set_allowance(&escrow_account_id, allowance - locked_balance);
+        self.accounts.insert(&owner_id, &account);
+    }
+
+    /// Locks an additional `lock_amount` to the caller of the function (`predecessor_id`) from
+    /// the `owner_id`.
+    /// Requirements:
+    /// * The (`predecessor_id`) should have enough allowance or be the owner.
+    /// * The owner should have enough unlocked balance.
+    pub fn lock(&mut self, owner_id: AccountId, lock_amount: Balance) {
+        if lock_amount == 0 {
+            env::panic(b"Can't lock 0 tokens");
+        }
+        let escrow_account_id = env::predecessor_account_id();
+        let mut account = self.get_account(&owner_id);
+
+        // Checking and updating unlocked balance
+        if account.balance < lock_amount {
+            env::panic(b"Not enough unlocked balance");
+        }
+        account.balance -= lock_amount;
+
+        // If locking by escrow, need to check and update the allowance.
+        if escrow_account_id != owner_id {
+            let allowance = account.get_allowance(&escrow_account_id);
+            if allowance < lock_amount {
+                env::panic(b"Not enough allowance");
```

Comment:
i have a question here, if paniced here the before `account.balance -= lock_amount` won't commit?

Reply:
Yes